### PR TITLE
Add graph ql docs

### PR DESCRIPTION
**Pull request by aklish**
Don't merge until elide 4 is released.

### CODE REVIEW

File: pages/guide/09-clientapis.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+layout: guide
+group: guide
+title: Client APIs
+---
+
+# API Usage (Client-side info)
+
+Graph APIs are an evolution of web service APIs that serve and manipulate data for mobile & web applications.
+They have a number of characteristics that make them well suited to this task:
+1.  Most notably, they present a **data model** as an entity relationship graph and **accompanying schema**.
```
Comment (DennisMcWherter) [inline pages/guide/09-clientapis.md]: insert: `an` before `accompanying schema`.

File: pages/guide/09-clientapis.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    * A well defined model allows for a consistent view of the data and a centralized way to manipulate an instance of the model or to cache it.
+    * The schema provides powerful introspection capabilities that can be used to build tools to help developers understand and navigate the model.
+2.  The API allows the client to **fetch or mutate as much or as little information in single roundtrip** between client and server.  This also
+    shrinks payload sizes and simplifies the process of schema evolution.
+3.  There is a **well defined standard** for the API that fosters a community approach to development of supporting tools & best practices.
+
+Elide supports the two most widely adopted standards for graph APIs: 
+
+* [JSON-API]({{site.baseurl}}/pages/guide/10-jsonapi.html)
```
Comment (DennisMcWherter) [inline pages/guide/09-clientapis.md:18]: idr, but I think we should have a better/more dynamic way of linking to other pages (to avoid broken links in the future if pages are moved). We should investigate.

File: pages/guide/10-jsonapi.md [hunk 3]
```diff
@@ -4,8 +4,20 @@ group: guide
...
+# JSON-API 
+
+[JSON-API](jsonapi.org) is a specification for building REST APIs for CRUD (create, read, update, and delete) operations.  
+Similar to graphQL: 
+*  It allows the client to control what is returned in the response payload.  
+*  It also offered an API extension (the _patch extension_) that allowed multiple mutations to the graph to occur in a single request.
+
+Unlike graphQL, it is more structured - laying out how to perform common API operations.  
+
+Also unlike graphQL, it has no standardized API introspection.  However, Elide adds this capability to any service by exporting
```
Comment (DennisMcWherter) [inline pages/guide/10-jsonapi.md]: Remove the `Also unlike graphQL`. Also, I'm not sure we're discussing "API" introspection (that's what the spec is for, right?) but instead, we're referring to the data's _schema_.

Would the following be a bit more natural?

`JSON-API has no standardized schema introspection.`

File: pages/guide/10-jsonapi.md [hunk 4]
```diff
@@ -4,8 +4,20 @@ group: guide
...
-In some scenarios, JSON-API does not define how an implementation should behave.  Elide also may not implement all optional features of the
-JSON-API specification.  The following sections clarify Elide's behavior with respect to undefined or optional aspects of JSON-API.
+# JSON-API 
+
+[JSON-API](jsonapi.org) is a specification for building REST APIs for CRUD (create, read, update, and delete) operations.  
+Similar to graphQL: 
+*  It allows the client to control what is returned in the response payload.  
+*  It also offered an API extension (the _patch extension_) that allowed multiple mutations to the graph to occur in a single request.
+
+Unlike graphQL, it is more structured - laying out how to perform common API operations.
```
Comment (DennisMcWherter) [inline pages/guide/10-jsonapi.md]: I think we should reword this to be a bit clearer. Maybe expand to something along the lines of this:

```
Unlike graphQL, JSON-API provides a rigorous structure
to its queries; it details the exact behavior of all common
API operations.
```

File: pages/guide/10-jsonapi.md [hunk 5]
```diff
@@ -16,34 +28,255 @@ There are a few caveats given that Elide allows developers control over how enti
...
 1. The root path segment of URLs are by default the name of the class (lowercase).  This can be overridden.
 1. Elide allows relationships to be nested arbitrarily deep in URLs.
-1. Elide currently requires all individual entities to be addressed by ID within a URL.  For example, consider a model with an article with a singular author which has a singular address.   While unambiguous, the following is *not* allowed: `/articles/1/author/address`.  Instead, the author must be fully qualified by ID: `/articles/1/author/34/address`
-
-## Filters
-
-Filters are covered in this [section]({{site.baseurl}}/pages/guide/08-filters.html).
-
+1. Elide currently requires all individual entities to be addressed by ID within a URL.  For example, consider a model with an 
+article with a singular author which has a singular address.   While unambiguous, the following is *not* allowed: `/articles/1/author/address`.
```
Comment (DennisMcWherter) [inline pages/guide/10-jsonapi.md]: Replace the second `with` -> `and`

`article and a singular author which has a singular address`

File: pages/guide/10-jsonapi.md [hunk 6]
```diff
@@ -16,34 +28,255 @@ There are a few caveats given that Elide allows developers control over how enti
...
+Instead, the author must be fully qualified by ID: `/articles/1/author/34/address`
 
 ## Model Identifiers
 
 Elide supports three mechanisms by which a newly created entity is assigned an ID:
 
 1. The ID is assigned by the client and saved in the data store.
 1. The client doesn't provide an ID and the data store generates one.
-1. The client provides an ID which is replaced by one generated by the data store.  When using the patch extension, the client
+1. The client provides an ID which is replaced by one generated by the data store.  When using _UPSERT_, the client
```
Comment (DennisMcWherter) [inline pages/guide/10-jsonapi.md]: JSON-API does not have UPSERT.

File: pages/guide/10-jsonapi.md [hunk 7]
```diff
@@ -16,34 +28,255 @@ There are a few caveats given that Elide allows developers control over how enti
...
-
 Elide looks for the JPA `GeneratedValue` annotation to disambiguate whether or not
 the data store generates an ID for a given data model.   If the client also generated 
 an ID during the object creation request, the data store ID overrides the client value.
 
 ### Matching newly created objects to IDs
 
-When using the patch extension, Elide always returns object entity bodies (containing newly assigned IDs) in 
-the order in which they were created.  The client can use this order to map the object created to its ID.
+When using _UPSERT_, Elide returns object entity bodies (containing newly assigned IDs) in
```
Comment (DennisMcWherter) [inline pages/guide/10-jsonapi.md]: No UPSERT in JSON-API.

File: pages/guide/10-jsonapi.md [hunk 8]
```diff
@@ -16,34 +28,255 @@ There are a few caveats given that Elide allows developers control over how enti
...
+
+JSON-API 1.0 is agnostic to filtering strategies.  The only recommendation is that servers and clients _should_
+prefix filtering query parameters with the word 'filter'.
+
+Elide supports multiple filter dialects and the ability to add new ones to meet the needs of developers or to evolve
+the platform should JSON-API standardize them.
+
+### Supported Dialects
+
+Elide supports two primary dialects - [RSQL](#rsql) and [basic](#basic).
```
Comment (DennisMcWherter) [inline pages/guide/10-jsonapi.md]: Remove `basic`. As of Elide 4.x I think we want to deprecate this in favor of RSQL. It is more complicated to use and more limited than RSQL.

File: pages/guide/10-jsonapi.md [hunk 9]
```diff
@@ -16,34 +28,255 @@ There are a few caveats given that Elide allows developers control over how enti
...
+
+#### Type Coercion
+Values are type coerced into the appropriate primitive data type for the attribute filter.
+
+#### Multiple Filters
+When multiple filters are present for the same type, the filters are a logical ‘and’ for any collection with
+that type.   Filters are type independent: any filters for type 'A' do *not* modify the results 
+returned for a collection of type 'B'.
+
+#### Basic Filter Examples
```
Comment (DennisMcWherter) [inline pages/guide/10-jsonapi.md]: I don't think we want to recommend Basic filtering anymore. We should probably deprecate in Elide 4.0 with plans to remove in 5.0.

File: pages/guide/10-jsonapi.md [hunk 10]
```diff
@@ -4,8 +4,20 @@ group: guide
 title: Json API
 ---
 
-In some scenarios, JSON-API does not define how an implementation should behave.  Elide also may not implement all optional features of the
-JSON-API specification.  The following sections clarify Elide's behavior with respect to undefined or optional aspects of JSON-API.
+# JSON-API 
+
+[JSON-API](jsonapi.org) is a specification for building REST APIs for CRUD (create, read, update, and delete) operations.  
+Similar to graphQL:
```
Comment (DennisMcWherter) [inline pages/guide/10-jsonapi.md]: We should be consistent with our casing of `GraphQL`. http://graphql.org would suggest that `GraphQL` is the appropriate way of casing it instead of `graphQL`. We also use the former casing in the GraphQL document below.

File: pages/guide/11-graphql.md [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+1.  **Query objects** which are used to compose queries and mutations
+2.  **Input Objects** which are used to supply input data to mutations
+
+The schema for both kinds of objects are derived from the entity relationship graph (defined by the JPA data model).
+Both contain a set of attributes and relationships.  Attributes are properties of the entity.
+Relationships are links to other entities in the graph.
+
+### Input Objects
+
+Input objects just contain attributes and relationship with names that directly match
```
Comment (DennisMcWherter) [inline pages/guide/11-graphql.md]: Remove `just`.

`match` -> `matching`

File: pages/guide/11-graphql.md [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+### Input Objects
+
+Input objects just contain attributes and relationship with names that directly match 
+the property names in the JPA annotated model:
+
+![GraphQL Input Object UML](/assets/images/graphql_input_object_uml.png){:class="img-responsive"}
+
+### Query Objects
+
+Query Objects are more complex given that queries have to support filtering, sorting,
```
Comment (DennisMcWherter) [inline pages/guide/11-graphql.md]: Reword?

```
Query Objects are more complex than Input Objects since they
do more than simply describe data; they must support filtering,
sorting, and pagination.
```

File: pages/guide/11-graphql.md [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+and pagination and not simply describe data.  Elide's GraphQL structure for queries and mutations is depicted below:
+
+![GraphQL Query Object UML](/assets/images/graphql_query_object_uml.png){:class="img-responsive"}
+
+Every GraphQL schema must define a root document which represents the root of the graph.
+In Elide, entities can be marked if they are directly navigable from the root of the
+graph. Elide’s GraphQL root documents consists of _relationships_ to these rootable entities.
```
Comment (DennisMcWherter) [inline pages/guide/11-graphql.md]: `consists` -> `consist`.

File: pages/guide/11-graphql.md [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+Each relationship is named by its pluralized type name in the GraphQL root document.
```
Comment (DennisMcWherter) [inline pages/guide/11-graphql.md]: `pluralized`? I don't think this is correct.

Each relationship derives its names from the relationship _field name_ which are often pluralized by humans.

File: pages/guide/11-graphql.md [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+
+All other non-rootable entities in our schema must be referenced through traversal of the
+relationships in the entity relationship graph.
+
+Every relationship is modeled the same way.  Elide adopts [Relay's pattern for pagination support](http://graphql.org/learn/pagination/).
```
Comment (DennisMcWherter) [inline pages/guide/11-graphql.md]: The way this jumps from the first sentence to the next is a bit confusing to me. I think if we reword the second sentence to something like

`Since Elide adopts ..., relationships are realized as a collection of _edges_ in practice.`

it may be clearer. Basically, at this moment (before continuing) I am confused on how we jumped from discussing our schema structure to Elide's adoption of Relay's pagination pattern.

File: pages/guide/11-graphql.md [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+Relationships are a collection of _edges_.  Each edge contains a _node_.  The _node_ is an instance of a
```
Comment (DennisMcWherter) [inline pages/guide/11-graphql.md]: I know what we're trying to explain here since I'm familiar with the system, but these final few sentences are really hard for me to parse for some reason. We should try to rework these.

File: pages/guide/11-graphql.md [hunk 17]
```diff
@@ -0,0 +1,277 @@
...
+data model that is also a member of the relationship.   It contains attributes and its own set of relationships.
+
+#### Relationship Arguments
+
+In GraphQL, any property in the schema can take arguments.  Relationships in Elide have a standard
+set of arguments that either constrain the edges fetched from a relationship or supply data to a mutation:
+
+1. The **ids** parameter is a collection of node identifiers.  It is used to select one or more nodes from a relationship.
+2. The **filter** parameter is used to build complex [RSQL](https://github.com/jirutka/rsql-parser) filter predicates that select zero or more nodes from a relationship.
```
Comment (DennisMcWherter) [inline pages/guide/11-graphql.md]: Remove `complex`?

File: pages/guide/11-graphql.md [hunk 18]
```diff
@@ -0,0 +1,277 @@
...
+3. The **sort** parameter is used to order a relationship's edges by one or more node attributes.
+4. The parameters **offset** and **first** are used to paginate a relationship across multiple API requests.
+5. The **op** argument describes the operation to perform on the relationship. When not provided, this argument
+defaults to a FETCH operation—which simply reads the collection of edges.
```
Comment (DennisMcWherter) [inline pages/guide/11-graphql.md]: Remove `-`

File: pages/guide/11-graphql.md [hunk 19]
```diff
@@ -0,0 +1,277 @@
...
+6. The **data** parameter is provided for operations that mutate the collection (UPSERT and REPLACE), It contains
+a list of input objects that match the data type of the relationship.
+
+Entity attributes generally do not take arguments. However, attributes can be annotated as computed which allows
```
Comment (DennisMcWherter) [inline pages/guide/11-graphql.md]: Computed attributes with arguments are not yet supported. Slated for 4.1, I believe.

File: pages/guide/11-graphql.md [hunk 20]
```diff
@@ -0,0 +1,277 @@
...
+**REPLACE** can be though of an **UPSERT** followed by an implicit **REMOVE** of everything else that was previously in the collection that the client
+has authorization to see & manipulate.
+
+#### Map Data Types
+
+GraphQL has no native support for a map data type.  If a JPA data model includes a map, Elide translates this to a list of key/value pairs in the GraphQL schema.
+
+### Making Calls
+
+All calls must be HTTP `POST` requests made to the root endpoint. This specific endpoint will depend on where you mount the provided servlet.
```
Comment (DennisMcWherter) [inline pages/guide/11-graphql.md]: This is true today, but general comment: we have an option to trivially support `GET` requests in the future if necessary.

File: pages/guide/11-graphql.md [hunk 21]
```diff
@@ -0,0 +1,277 @@
...
+POST https://yourdomain.com/graphql
+```
+
+### Example Data Model
+
+All subsequent query examples are based on the following data model including `Book`, `Author`, and `Publisher`:
+
+{% include code_example example='graphql-data-model' offset=0 %}
+
+### Filtering
```
Comment (DennisMcWherter) [inline pages/guide/11-graphql.md]: This is duplication. We should instead outline these sections and refer to the _specific_ sections for the relevant topics.

If, however, we really want to repeat the information here (I don't have strong opinions on this), we should at least be as DRY as possible for information duplication. We should just have a _single_ source of truth for this information and it should get imported by jekyll here.

**Review by DennisMcWherter — CHANGES_REQUESTED** [review]

File: pages/guide/09-clientapis.md (hunk 2 above)
Comment (aklish) [reply pages/guide/09-clientapis.md:18]: Agreed.  Had the same thoughts.

File: pages/guide/11-graphql.md (hunk 14 above)
Comment (aklish) [reply pages/guide/11-graphql.md]: I'm only referring to the root document relationships.

File: pages/guide/11-graphql.md (hunk 20 above)
Comment (aklish) [reply pages/guide/11-graphql.md]: We'll document future stuff when it exists I guess.

File: pages/guide/11-graphql.md (hunk 21 above)
Comment (aklish) [reply pages/guide/11-graphql.md]: Agreed but not sure the best way to fix this now given the new document structure.

File: pages/guide/11-graphql.md [hunk 22]
```diff
@@ -0,0 +1,308 @@
...
+
+1. The **FETCH** operation retrieves a set of objects. When a list of ids is specified, it will only extract the set of objects within the
+relationship with matching ids.  If no ids are specified, then the entire collection of objects will be returned to the caller.
+2. The **DELETE** operation fully deletes an object from the system.
+3. The **REMOVE** operation removes a specified set (qualified by the _ids_ argument) of objects from a relationship. This allows the caller to remove
+relationships between objects without being forced to fully delete the referenced objects.
+4. The **UPSERT** operation behaves much like SQL’s MERGE.  Namely, if the object already exists (based on the provided
+id) then it will be updated.  Otherwise, it will be created. In the case of updates, attributes that are not specified are left unmodified.
+5. The **REPLACE** operation is intended to replace an entire relationship with the set of objects provided in the _data_ argument.
+**REPLACE** can be though of an **UPSERT** followed by an implicit **REMOVE** of everything else that was previously in the collection that the client
```
Comment (wcekan) [inline pages/guide/11-graphql.md]: can be `thought` of `as` an?

File: pages/guide/10-jsonapi.md [hunk 23]
```diff
@@ -33,17 +43,211 @@ Elide supports three mechanisms by which a newly created entity is assigned an I
...
+JSON-API 1.0 is agnostic to filtering strategies.  The only recommendation is that servers and clients _should_
+prefix filtering query parameters with the word 'filter'.
+
+Elide supports multiple filter dialects and the ability to add new ones to meet the needs of developers or to evolve
+the platform should JSON-API standardize them.  Elide's primary dialect is [RSQL](https://github.com/jirutka/rsql-parser)
+
+### RSQL
+
+[RSQL](https://github.com/jirutka/rsql-parser) is a query language that allows conjunction (and), disjunction (or), and parenthetic grouping
+of boolean expressions.  It is a superset of the [FIQL language](https://tools.ietf.org/html/draft-nottingham-atompub-fiql-00).
```
Comment (wcekan) [inline pages/guide/10-jsonapi.md]: Should be capitalized as `Boolean` since it is a name.

File: pages/guide/11-graphql.md [hunk 24]
```diff
@@ -0,0 +1,308 @@
...
+* `=lt=` : Evaluates to true if the attribute is less than the value.
+* `=gt=` : Evaluates to true if the attribute is greater than the value.
+* `=le=` : Evaluates to true if the attribute is less than or equal to the value.
+* `=ge=` : Evaluates to true if the attribute is greater than or equal to the value.
+
+### Examples
+* Filter books by title equal to 'abc' _and_ genre starting with 'Science':
+  `"title=='abc';genre=='Science*'` 
+* Filter books with a publication date greater than a certain time _or_ the genre is _not_ 'Literary Fiction'
+or 'Sicence Fiction':
```
Comment (wcekan) [inline pages/guide/11-graphql.md]: `Science`

File: pages/guide/11-graphql.md [hunk 25]
```diff
@@ -0,0 +1,308 @@
...
+Any relationship can be sorted by attributes in:
+* The relationship model
+* Another model joined to the relationship model through to-one relationships
+
+To join across relationships, the attribute name is prefixed by one or more relationship names separated by period ('.')
+
+It is also possible to sort in either ascending or descending order by prepending
+the attribute expression with a '+' or '-' character.  If no order character is provided, sort order defaults to ascending.
+
+A relationship can be sorted by multiple attributes by seperating the attribute expressions by commas: ','.
```
Comment (wcekan) [inline pages/guide/11-graphql.md]: separating
